### PR TITLE
WIP: Major rework of the openqa bot 

### PR DESCRIPTION
**Pull request by coolo**
I'm working on this for a week and more - it's no longer looking at requests, but at incident projects and test updates and then only aggregates into the request if it opened.

This way the release managers get to know about problems in incident *before* they open the request into testing.

But I didn't find a good way to make this work for opensuse too - and the kgraft testing needs work too.

### CODE REVIEW

**Comment by coolo** [comment]
Hmm, the api is a bit above my head right now. Perhaps coffee helps

File: openqa-maintenance.py [hunk 1]
```diff
@@ -91,73 +96,191 @@
 with open(opa.join(data_path, "data/repos.json"), 'r') as f:
     TARGET_REPO_SETTINGS = json.load(f)
 
+with open(opa.join(data_path, "data/apimap.json"), 'r') as f:
+    API_MAP = json.load(f)
+
+MINIMALS = dict()
```
Comment (mimi1vx) [inline openqa-maintenance.py]: this should be 'set'

File: openqa-maintenance.py [hunk 2]
```diff
@@ -91,73 +96,191 @@
...
+        root = ET.fromstring(repomd.text)
+
+        cs = root.find(
+            './/{http://linux.duke.edu/metadata/repo}data[@type="updateinfo"]/{http://linux.duke.edu/metadata/repo}location')
+        url = repo + cs.attrib['href']
+
+        # python 3 brings gzip.decompress, but with python 2 we need to store to
+        # temporary file to uncompress
+        repomd = requests.get(url).content
+        tfile = NamedTemporaryFile()
```
Comment (mimi1vx) [inline openqa-maintenance.py]: whole python3 magic :

```
def decompress(data):
    """Decompress a gzip compressed string in one shot.
    Return the decompressed string.
    """
    with GzipFile(fileobj=io.BytesIO(data)) as f:
        return f.read()
```

**Comment by jberry-suse** [comment]
hmm, assumed this was going to be changed to non-WIP in title once ready for final review and perhaps commits like "one more" changed or squashed. Rather messy to merge this as is, that's why git-rebase exists. :(

**Comment by jberry-suse** [comment]
Also looks like the one comment I made regarding using the comment code that we now have rather than adding/maintaining yet another copy was also left unchanged...

File: openqa-maintenance.py [hunk 3]
```diff
@@ -927,9 +969,9 @@ def test(self):
                 result = 'declined'
                 state = 'done'
             comment = "<!-- openqa state=%s result=%s revision=%s -->\n" % (state, result, job['base'].get('openqa_build'))
-            comment += "\n" + msg
+            comment += "\nCC @coolo\n" + msg
 
-            #print comment
+            # print comment
```
Comment (jberry-suse) [inline openqa-maintenance.py]: huh?

File: openqa-maintenance.py [hunk 4]
```diff
@@ -915,7 +957,7 @@ def test(self):
                     openqa_jobs += jobs
             if not openqa_done or len(openqa_jobs) == 0:
                 continue
-            #print openqa_jobs
+            # print openqa_jobs
```
Comment (jberry-suse) [inline openqa-maintenance.py]: ??

File: openqa-maintenance.py [hunk 5]
```diff
@@ -879,34 +920,35 @@ def check_product(self, job, product_prefix):
...
         for inc in requests.get('https://maintenance.suse.de/api/incident/active/').json():
-            if not inc in ['a4871', 'a5146', 'a2129', '5219', '5217', '5230']: continue
-            #if not inc.startswith('52'): continue
+            if not inc in ['a4871', 'a5205', 'a2129', 'a5219', '5258', '5223']:
+                continue
+            if not inc.startswith('52'):
+                continue
             print inc
-            #continue
+            # continue
```
Comment (jberry-suse) [inline openqa-maintenance.py]: ??

File: openqa-maintenance.py [hunk 6]
```diff
@@ -269,10 +288,22 @@ def kernel_target(req):
...
+        if not minimal:
+            return []
+
+        settings = settings.copy()
+        settings['FLAVOR'] += '-Minimal'
+        return [settings]
+
+    def add_kernel_settings(self, prj, settings):
+        settings = settings.copy()
+        # not right now
```
Comment (jberry-suse) [inline openqa-maintenance.py]: indeed, that's what `return` means.

File: openqa-maintenance.py [hunk 7]
```diff
@@ -792,15 +841,119 @@ def check_one_request(self, req):
...
+        if comment_info.get('comment', '') == comment:
+            self.logger.debug("%s comment did not change", incident_project)
+            return
+
+        self.logger.debug("adding comment to %s, state %s", incident_project, state)
+        #self.logger.debug("message: %s", msg)
+        if not self.dryrun:
+            if comment_id is not None:
+                self.commentapi.delete(comment_id)
+            self.commentapi.add_comment(project_name=str(incident_project), comment=str(comment))
```
Comment (jberry-suse) [inline openqa-maintenance.py]: Virtually all of this is generalized in `ReviewBot`....

**Comment by mimi1vx** [comment]
@jberry-suse, in reality, this code was used in production for more than a month :( so it mainly synncs with reality

**Comment by jberry-suse** [comment]
I did not comment on that and do not doubt it. Just asking that we raise the quality bar, just a tad. If I didn't rebase my garbage commits this repo would have 10k useless and unintelligible commits from me.

**Comment by jberry-suse** [comment]
I do the same myself, but that does not prevent me from going back and reworking all my commits to actually make sense and have useful messages instead of two undescriptive words.

**Comment by coolo** [comment]
just realized Ondrej merged :(

This was never meant to be merged as is - this was supposed to be reviewed and tested and improved.

**Pull request merged**